Compiler crash with overriding type members in JS types

### Repro testcase here https://github.com/helllamer/sjs140ntse-testcase
Steps to reproduce:
1. `sbt`
2. `compile`


```
sbt:sjs140ntse> compile
[warn] There may be incompatibilities among your library dependencies; run 'evicted' to see detailed eviction warnings.
[info] compiling 2 Scala sources to sjs140ntse/target/scala-2.13/classes ...
[error] ## Exception when compiling 2 sources to sjs140ntse/target/scala-2.13/classes
[error] java.lang.AssertionError: assertion failed: 
[error]   got non-term symbol: type Element_t
[error]      while compiling: sjs140ntse/src/main/scala/react/leaflet/lmap/LocateControlR.scala
[error]         during phase: jsinterop
[error]      library version: version 2.13.4
[error]     compiler version: version 2.13.4
.......
[error] 
[error]   last tree to typer: Select(Select(Ident(scala), scalajs), js)
[error]        tree position: line 25 of sjs140ntse/src/main/scala/react/leaflet/lmap/MapComponentR.scala
[error]             tree tpe: scala.scalajs.js.type
[error]               symbol: final package js in package scalajs
[error]    symbol definition: final package js (a ModuleSymbol)
[error]       symbol package: scala.scalajs
[error]        symbol owners: package js
[error]            call site: package <root> in <none>
[error] 
[error] == Source file context for tree position ==
[error] 
[error]     22 }
[error]     23 
[error]     24 
[error]     25 @js.native
[error]     26 trait Context extends js.Object {
[error]     27 }
[error]     28 
[error] Total time: 1 s, completed 15.01.2021 12:45:54
```



Issue related to `override type Element_t` member in js-subclass:
```
@JSImport("react-leaflet", "MapComponent")
@js.native
class MapComponentR(...) extends js.Object {
  type Element_t <: js.Object
  var leafletElement: Element_t = js.native
}
```
```
sealed class LocateControlC(_props: LocateControlOptions, _ctx: Context)
  extends MapComponentR[LocateControlOptions](_props, _ctx)
{
  override type Element_t = LocateControl
  def componentWillMount(): Unit = {
    leafletElement = new LocateControl(props)
  }
}
```

<!--
Please follow these steps to help us fixing the bug:

1. Make sure it is a bug for this repo
  * Can you reproduce the bug with Scala/JVM? If yes, report it to https://github.com/scala/bug instead
  * Make sure it is not one of the intended semantic differences: https://www.scala-js.org/doc/semantics.html
  * Does the bug involve macros? If yes, make sure to reproduce *without* macros, or file an issue to the relevant macro library instead or ask on Gitter if it's your own macro
  * Is it about the website? If yes, report it to https://github.com/scala-js/scala-js-website/issues instead
  * If in doubt, ask on Gitter first: https://gitter.im/scala-js/scala-js

2. Make sure we can understand and reproduce the bug
  * Give code *and* reproducible steps to provoke the bug
  * Give the version of Scala and Scala.js you are using (if relevant, sbt as well)
  * What did you expect to happen?
  * What actually happened?

If possible, minimize the bug. This helps fixing bugs more quickly. Try to get rid of any external library.

Thank you!
-->

--- a/compiler/src/main/scala/org/scalajs/nscplugin/PrepJSInterop.scala
+++ b/compiler/src/main/scala/org/scalajs/nscplugin/PrepJSInterop.scala
@@ -681,46 +681,56 @@ abstract class PrepJSInterop[G <: Global with Singleton](val global: G)
         val low = overridingPair.low
         val high = overridingPair.high
 
-        def errorPos = {
-          if (sym == low.owner) low.pos
-          else if (sym == high.owner) high.pos
-          else sym.pos
-        }
+        if (low.isType || high.isType) {
+          /* #4375 Do nothing if either is a type, and let refchecks take care
+           * of it.
+           * The case where one is a type and the other is not should never
+           * happen, because they would live in different namespaces and
+           * therefore not override each other. However, if that should still
+           * happen for some reason, rechecks should take care of it as well.
+           */
+        } else {
+          def errorPos = {
+            if (sym == low.owner) low.pos
+            else if (sym == high.owner) high.pos
+            else sym.pos
+          }
 
-        def memberDefString(membSym: Symbol): String =
-          membSym.defStringSeenAs(sym.thisType.memberType(membSym))
+          def memberDefString(membSym: Symbol): String =
+            membSym.defStringSeenAs(sym.thisType.memberType(membSym))
 
-        // Check for overrides with different JS names - issue #1983
-        if (jsInterop.JSCallingConvention.of(low) != jsInterop.JSCallingConvention.of(high)) {
-          val msg = {
-            def memberDefStringWithCallingConvention(membSym: Symbol) = {
-              memberDefString(membSym) +
-              membSym.locationString + " called from JS as " +
-              JSCallingConvention.of(membSym).displayName
+          // Check for overrides with different JS names - issue #1983
+          if (jsInterop.JSCallingConvention.of(low) != jsInterop.JSCallingConvention.of(high)) {
+            val msg = {
+              def memberDefStringWithCallingConvention(membSym: Symbol) = {
+                memberDefString(membSym) +
+                membSym.locationString + " called from JS as " +
+                JSCallingConvention.of(membSym).displayName
+              }
+              "A member of a JS class is overriding another member with a different JS calling convention.\n\n" +
+              memberDefStringWithCallingConvention(low) + "\n" +
+              "    is conflicting with\n" +
+              memberDefStringWithCallingConvention(high) + "\n"
             }
-            "A member of a JS class is overriding another member with a different JS calling convention.\n\n" +
-            memberDefStringWithCallingConvention(low) + "\n" +
-            "    is conflicting with\n" +
-            memberDefStringWithCallingConvention(high) + "\n"
+
+            reporter.error(errorPos, msg)
           }
 
-          reporter.error(errorPos, msg)
-        }
+          /* Cannot override a non-@JSOptional with an @JSOptional. Unfortunately
+           * at this point the symbols do not have @JSOptional yet, so we need
+           * to detect whether it would be applied.
+           */
+          if (!isJSNative) {
+            def isJSOptional(sym: Symbol): Boolean = {
+              sym.owner.isTrait && !sym.isDeferred && !sym.isConstructor &&
+              !sym.owner.hasAnnotation(JSNativeAnnotation)
+            }
 
-        /* Cannot override a non-@JSOptional with an @JSOptional. Unfortunately
-         * at this point the symbols do not have @JSOptional yet, so we need
-         * to detect whether it would be applied.
-         */
-        if (!isJSNative) {
-          def isJSOptional(sym: Symbol): Boolean = {
-            sym.owner.isTrait && !sym.isDeferred && !sym.isConstructor &&
-            !sym.owner.hasAnnotation(JSNativeAnnotation)
-          }
-
-          if (isJSOptional(low) && !(high.isDeferred || isJSOptional(high))) {
-            reporter.error(errorPos,
-                s"Cannot override concrete ${memberDefString(high)} from " +
-                s"${high.owner.fullName} in a non-native JS trait.")
+            if (isJSOptional(low) && !(high.isDeferred || isJSOptional(high))) {
+              reporter.error(errorPos,
+                  s"Cannot override concrete ${memberDefString(high)} from " +
+                  s"${high.owner.fullName} in a non-native JS trait.")
+            }
           }
         }
       }

--- a/compiler/src/test/scala/org/scalajs/nscplugin/test/JSInteropTest.scala
+++ b/compiler/src/test/scala/org/scalajs/nscplugin/test/JSInteropTest.scala
@@ -2991,6 +2991,25 @@ class JSInteropTest extends DirectTest with TestHelpers {
     }
     """.hasNoWarns()
 
+    // #4375
+    """
+    abstract class Parent extends js.Object {
+      type TypeMember <: CharSequence
+      type JSTypeMember <: js.Object
+
+      type Foo = Int
+      @JSName("Babar") def Bar: Int = 5
+    }
+
+    class Child extends Parent {
+      type TypeMember = String
+      override type JSTypeMember = js.Date // the override keyword makes no difference
+
+      @JSName("Foobar") def Foo: Int = 5
+      type Bar = Int
+    }
+    """.hasNoWarns()
+
     """
     abstract class A extends js.Object {
       @JSName("foo")

